### PR TITLE
LCO Telescopes - Sinistro, Spectral, Muscat, Floyds APIs

### PR DESCRIPTION
**Pull request by mcoughlin**
LCO triggering interface. No tests currently committed as they require existing program to test (no response from LCO on powerless account).

### CODE REVIEW

**Comment by mcoughlin** [comment]
Maybe @dannygoldstein @kmshin1397 can review this like LT?

**Comment by dannygoldstein** [comment]
@mcoughlin Sure, ping us when ready.

**Comment by mcoughlin** [comment]
@dannygoldstein this version is fine, but we need to discuss what you would like done about tests, if any.

**Comment by dannygoldstein** [comment]
@mcoughlin - can you add these to demo data so that I can play around with the rendered forms?

**Comment by mcoughlin** [comment]
@dannygoldstein You mean without the keys required to actually submit? Or do you want I send you those too separately?

**Comment by dannygoldstein** [comment]
Hm, didn't think about that. Is there a test account at LCO?

**Comment by mcoughlin** [comment]
@dannygoldstein no

**Comment by dannygoldstein** [comment]
OK. Can we get one going? Will need one for tests and debugging.

**Comment by mcoughlin** [comment]
@dannygoldstein they showed no interest in making such an account.

**Comment by dannygoldstein** [comment]
I think we'll need one to move forward, or at least some way of validating that dummy requests get the expected output.

**Comment by mcoughlin** [comment]
@dannygoldstein it would seem a bit disappointing if we are going to delay getting a key piece of our follow-up ecosystem into skyportal for this. can we discuss alternatives?

**Comment by dannygoldstein** [comment]
@mcoughlin: @kmshin1397  will also need the creds

**Comment by dannygoldstein** [comment]
@mcoughlin 

When I try to submit a request to SPECTRAL locally, I get:

```
Error: Bad Request for url: https://observe.lco.global/api/requestgroups/
```

**Comment by mcoughlin** [comment]
@dannygoldstein Is it in the right hemisphere?

**Comment by dannygoldstein** [comment]
Not sure - is there a way to introspect the error message provided by LCO? (I.e., "this request failed because ... ")

**Comment by dannygoldstein** [comment]
object was 16fil -  21:30:52.53  +27:34:26.81

**Comment by mcoughlin** [comment]
@dannygoldstein ok I pushed something like that now.. let's see.

File: skyportal/facility_apis/lco.py [hunk 1]
```diff
@@ -0,0 +1,1032 @@
...
+            'type': 'ICRS',
+            'ra': request.obj.ra,
+            'dec': request.obj.dec,
+            'epoch': 2000,
+        }
+
+        exp_time = request.payload["exposure_time"]
+        exp_count = int(request.payload["exposure_counts"])
+
+        # The configurations for this request. In this example we are taking 2 exposures with different filters.
```
Comment (kmshin1397) [inline skyportal/facility_apis/lco.py:63]: The comment about a 2-exposure example is no longer relevant, no?

File: skyportal/facility_apis/lco.py [hunk 2]
```diff
@@ -0,0 +1,1032 @@
...
+        if not altdata:
+            return
+
+        content = req.transactions[0].response["content"]
+        content = json.loads(content)
+        uid = content["id"]
+
+        r = requests.post(
+            'https://observe.lco.global/api/requestgroups/{}/cancel/'.format(uid),
+            headers={'Authorization': 'Token {}'.format(altdata["API_TOKEN"])},
```
Comment (kmshin1397) [inline skyportal/facility_apis/lco.py:518]:
```suggestion
            f"https://observe.lco.global/api/requestgroups/{uid}/cancel/",
            headers={"Authorization": f'Token {altdata["API_TOKEN"]}'},
```
I think using the f formatting is preferable to format here for consistency with the rest of the codebase

File: skyportal/facility_apis/lco.py [hunk 3]
```diff
@@ -0,0 +1,1032 @@
...
+        if not altdata:
+            return
+
+        content = req.transactions[0].response["content"]
+        content = json.loads(content)
+        uid = content["id"]
+
+        r = requests.get(
+            'https://observe.lco.global/api/requestgroups/{}/'.format(uid),
+            headers={'Authorization': 'Token {}'.format(altdata["API_TOKEN"])},
```
Comment (kmshin1397) [inline skyportal/facility_apis/lco.py:572]:
```suggestion
            f"https://observe.lco.global/api/requestgroups/{uid}/",
            headers={"Authorization": f'Token {altdata["API_TOKEN"]}'},
```

File: skyportal/facility_apis/lco.py [hunk 4]
```diff
@@ -0,0 +1,1032 @@
...
+        altdata = request.allocation.altdata
+        if not altdata:
+            return
+
+        lcoreq = SINISTRORequest(request)
+        requestgroup = lcoreq.requestgroup
+
+        r = requests.post(
+            'https://observe.lco.global/api/requestgroups/',
+            headers={'Authorization': 'Token {}'.format(altdata["API_TOKEN"])},
```
Comment (kmshin1397) [inline skyportal/facility_apis/lco.py:620]:
```suggestion
            "https://observe.lco.global/api/requestgroups/",
            headers={"Authorization": f'Token {altdata["API_TOKEN"]}'},
```

File: skyportal/facility_apis/lco.py [hunk 5]
```diff
@@ -0,0 +1,1032 @@
...
+
+        if not altdata:
+            return
+
+        lcoreq = SPECTRALRequest(request)
+        requestgroup = lcoreq.requestgroup
+
+        r = requests.post(
+            'https://observe.lco.global/api/requestgroups/',
+            headers={'Authorization': 'Token {}'.format(altdata["API_TOKEN"])},
```
Comment (kmshin1397) [inline skyportal/facility_apis/lco.py:734]:
```suggestion
            "https://observe.lco.global/api/requestgroups/",
            headers={"Authorization": f'Token {altdata["API_TOKEN"]}'},
```

File: skyportal/facility_apis/lco.py [hunk 6]
```diff
@@ -0,0 +1,1032 @@
...
+        altdata = request.allocation.altdata
+        if not altdata:
+            return
+
+        lcoreq = MUSCATRequest(request)
+        requestgroup = lcoreq.requestgroup
+
+        r = requests.post(
+            'https://observe.lco.global/api/requestgroups/',
+            headers={'Authorization': 'Token {}'.format(altdata["API_TOKEN"])},
```
Comment (kmshin1397) [inline skyportal/facility_apis/lco.py:850]:
```suggestion
            "https://observe.lco.global/api/requestgroups/",
            headers={"Authorization": f'Token {altdata["API_TOKEN"]}'},
```

File: skyportal/facility_apis/lco.py [hunk 7]
```diff
@@ -0,0 +1,1032 @@
...
+        altdata = request.allocation.altdata
+        if not altdata:
+            return
+
+        lcoreq = FLOYDSRequest(request)
+        requestgroup = lcoreq.requestgroup
+
+        r = requests.post(
+            'https://observe.lco.global/api/requestgroups/',
+            headers={'Authorization': 'Token {}'.format(altdata["API_TOKEN"])},
```
Comment (kmshin1397) [inline skyportal/facility_apis/lco.py:956]:
```suggestion
            "https://observe.lco.global/api/requestgroups/",
            headers={"Authorization": f'Token {altdata["API_TOKEN"]}'},
```

**Review by kmshin1397 — COMMENTED** [review]
I'm also seeing the SPECTRAL error that Danny mentioned:
image
Also, trying to delete the failed request from above also threw an error:
Screenshot 2021-01-14 093354

**Comment by mcoughlin** [comment]
@kmshin1397 So basically what is going on is I need a good way to ensure that for Spectral, requestgroup['requests']['target']["dec"] is less than 17 degrees. Its not quite the same as the min and max in the form, can you think of a good way to ensure this? I think throwing an error afterwards doesnt seem to be a good solution.

**Comment by kmshin1397** [comment]
> @kmshin1397 So basically what is going on is I need a good way to ensure that for Spectral, requestgroup['requests']['target']["dec"] is less than 17 degrees. Its not quite the same as the min and max in the form, can you think of a good way to ensure this? I think throwing an error afterwards doesnt seem to be a good solution.

Ah I see, yeah can't really use the form validation as it's taken from the obj itself. I would say the best approach would be to check for it and throw an error in the `_build_payload` functions so that the error occurs before the LCO API is actually hit, and we can send back a more informative message in that case.

**Comment by mcoughlin** [comment]
@kmshin1397 great idea. Implemented.

**Comment by kmshin1397** [comment]
> @kmshin1397 great idea. Implemented.

@mcoughlin Awesome. Looks like I lied earlier - FLOYDS seems fine to me now but still seeing MUSCAT fail. Is there a similar constraint for it we should apply?

**Comment by kmshin1397** [comment]
Also still need to add handling for failed submissions in the delete method I think. The list index error would indicate it's probably failing on the `content = req.transactions[0].response["content"]` line (lco.py:514). I guess since we failed the submission there is no transaction there. It would still be good to be able to remove the failed record from the list of requests.

**Comment by mcoughlin** [comment]
@kmshin1397 I have run out of time on this instrument in my program. Had to check my lco page to be sure.

**Comment by mcoughlin** [comment]
@dannygoldstein How do you feel about the failed deletes?

**Comment by dannygoldstein** [comment]
Same as the rest - we need to be able to test.

________________________________

**Comment by mcoughlin** [comment]
@kmshin1397 @dannygoldstein If I remove r.raise_for_status, I can get the tests to throw issues like the attached. Do we need the raise for status? 

Screen Shot 2021-01-16 at 3 59 54 PM

**Comment by mcoughlin** [comment]
@dannygoldstein So checking the validates step, it returns 
status_code == 200
instead of
status_code == 201
but otherwise seems like it can be made to imitate submission OK. Is there a good way to determine from
env, cfg = load_env() 
if we are testing or not? If so, we could do just some kind of if then to check with url to query?

**Comment by acrellin** [comment]
@dannygoldstein Any updates? Looks like @mcoughlin is waiting for feedback from you.

**Comment by mcoughlin** [comment]
@acrellin @dannygoldstein @kmshin1397 I think we will redo this test with Kyung to use his interface (submitting to an active program and then deleting). But we need to sit down together or I send my info or something.

**Pull request closed without merging**